omod/src/main/java/org/openmrs/module/spa/SpaController.java

```diff
@@ -0,0 +1,121 @@
+package org.openmrs.module.spa;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.core.io.FileSystemResource;
+import org.springframework.core.io.Resource;
+import org.springframework.http.CacheControl;
+import org.springframework.http.HttpRequest;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.net.URLConnection;
+import java.nio.file.Files;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+@RequestMapping(value = {"/spa", "/ws/spa"}, method = RequestMethod.GET)
+@Slf4j
+public class SpaController {
+
+    private static final Pattern URL_PATTERN = Pattern.compile("(?:/ws)?/spa");
+
+    private final SpaResourceLoader resourceLoader;
+
+    @Autowired
+    SpaController(@Qualifier("spaResourceLoader") SpaResourceLoader resourceLoader) {
+        this.resourceLoader = resourceLoader;
+    }
+
+    /**
+     * This controller method handles importmap.json, routes.registry.json and service-worker.js, ensuring that they are
+     * sent with no-cache and must-revalidate headers.
+     * <p/>
+     * Note that if you are using the {@code --hash-import-map} feature of the {@code assemble} command, your distro's
+     * importmap will look like importmap.&lt;hash&gt;.json and <em>should</em> be cached.
+     */
+    @RequestMapping("/{filename:(?:importmap|routes\\.registry)\\.json|service-worker\\.js}")
+    public ResponseEntity<Resource> getStaticFileWithoutCacheHeaders(@PathVariable String filename) {
+        FileSystemResource resource = (FileSystemResource) resourceLoader.getResource("/" + filename);
+        if (resource.exists()) {
+            MediaType mediaType = filename.equalsIgnoreCase("service-worker") ? MediaType.parseMediaType("text/javascript") : MediaType.APPLICATION_JSON;
```

Comment:
Wouldn't it be better to just look at the file extension here?

Reply:
Maybe? But this is a very specific rule to cover only three files by name, so actually parsing the extension seemed like overkill. (Most requests are handled by the function that does that; these are the files that cannot be cached without making the app non-functional).